crates/local-cluster-runner/src/node/mod.rs

```diff
@@ -0,0 +1,740 @@
+        }
+    }
+
+    pub fn lines(&self, pattern: Regex) -> impl Stream<Item = String> + '_ {
+        match self.status {
+            StartedNodeStatus::Exited { .. } => futures::stream::empty().left_stream(),
+            StartedNodeStatus::Failed { .. } => futures::stream::empty().left_stream(),
+            StartedNodeStatus::Running { ref searcher, .. } => {
+                let receiver = searcher.search(pattern);
+                receiver.right_stream()
+            }
+        }
+    }
+
+    pub async fn metadata_client(
+        &self,
+    ) -> Result<restate_metadata_store::MetadataStoreClient, GenericError> {
+        restate_metadata_store::local::create_client(
+            self.config.common.metadata_store_client.clone(),
+        )
+        .await
+    }
+
+    pub async fn admin_healthy(&self) -> bool {
+        if let Some(address) = self.admin_address() {
+            match reqwest::get(format!("http://{address}/health")).await {
+                Ok(resp) => resp.status().is_success(),
+                Err(_) => false,
+            }
+        } else {
+            false
+        }
+    }
+
+    pub async fn wait_admin_healthy(&self, timeout: Duration) -> bool {
+        let mut attempts = 1;
+        if tokio::time::timeout(timeout, async {
+            while !self.admin_healthy().await {
+                attempts += 1;
+                tokio::time::sleep(Duration::from_millis(250)).await
+            }
+        })
+        .await
+        .is_ok()
+        {
+            info!(
+                "Node {} is healthy after {attempts} attempts",
+                self.node_name(),
+            );
+            true
+        } else {
+            warn!(
+                "Timed out waiting for admin health on node {}",
+                self.node_name()
+            );
+            false
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Searcher {
+    inner: Arc<ArcSwapOption<SearcherInner>>,
+}
+
+#[derive(Clone)]
+struct SearcherInner {
+    set: RegexSet,
+    senders: Vec<Arc<Sender<String>>>,
+}
+
+impl Searcher {
+    fn new() -> Self {
+        Self {
+            inner: Arc::new(ArcSwapOption::from_pointee(SearcherInner {
+                set: RegexSet::empty(),
+                senders: Vec::new(),
+            })),
+        }
+    }
+
+    fn close(&self) {
+        self.inner.rcu(|_| None);
+    }
+
+    async fn matches(&self, haystack: &str) {
+        let inner = self.inner.load();
+        if let Some(inner) = &*inner {
+            for i in inner.set.matches(haystack) {
+                // ignore sender errors; it just means we have an outdated inner and the receiver
+                // has since dropped.
+                _ = inner.senders[i].send(haystack.to_owned()).await;
+            }
+        }
+    }
+
+    pub fn search(&self, regex: Regex) -> Receiver {
+        let (sender, receiver) = mpsc::channel(1);
+        let sender = Arc::new(sender);
+        let sender_ptr = Arc::as_ptr(&sender);
```

Comment:
Would `WeakSender` work in this case?

Reply:
yes, if i upgraded the weak sender then i could use the same_channel on sender. but this method just does an arc ptr comparison anyway, so im not sure if its worth going through the upgrade (which is fallible)